sway/commands/sandbox_socket.c

```diff
@@ -0,0 +1,119 @@
+#define _XOPEN_SOURCE 700 // for strdup
+#include <unistd.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <string.h>
+#include <strings.h>
+#include <sys/socket.h>
+#include <sys/un.h>
+#include <wayland-server-core.h>
+#include "sway/commands.h"
+#include "sway/security.h"
+#include "sway/tree/container.h"
+#include "sway/tree/view.h"
+#include "list.h"
+#include "log.h"
+#include "util.h"
+
+struct sandbox_socket {
+	char* path;
+	struct wl_event_source *src;
+	int fd;
+	uint32_t label_id;
+};
+
+static list_t *sandbox_sockets;
+
+static int fd_accept(int srv_fd, uint32_t mask, void *data) {
+	struct sandbox_socket *sock = data;
+	struct sockaddr_un name;
+	socklen_t slen = sizeof(name);
+
+	int cli_fd = accept(srv_fd, (struct sockaddr *)&name, &slen);
+	if (cli_fd < 0)
+		return 0;
+
+	sway_set_cloexec(cli_fd, true);
+	struct wl_client* client = wl_client_create(server.wl_display, cli_fd);
+	if (client) {
+		security_set_client_label_id(client, sock->label_id);
+	} else {
+		close(cli_fd);
+	}
+
+	return 1;
+}
+
+struct cmd_results *cmd_sandbox_socket(int argc, char **argv) {
+	struct cmd_results *error = NULL;
+	if ((error = checkarg(argc, "sandbox_socket", EXPECTED_AT_LEAST, 2))) {
+		return error;
+	}
+
+	if (!sandbox_sockets)
+		sandbox_sockets = create_list();
+
+	char* op = argv[0];
+	char* path = argv[1];
+
+	if (strcmp(op, "create") == 0) {
+		struct sockaddr_un name = {};
+		size_t path_len = strlen(path) + 1;
+		char* label = NULL;
+		int i = 2;
+		while (i < argc) {
+			if (strcmp(argv[i], "--label") == 0) {
+				if (i + 1 >= argc)
+					return cmd_results_new(CMD_INVALID, "--label requires an argument");
+				label = argv[i + 1];
+				i += 2;
+			} else {
+				return cmd_results_new(CMD_INVALID, "Unknown argument to sandbox_socket");
+			}
+		}
+
+		if (path_len > sizeof(name.sun_path)) {
+			return cmd_results_new(CMD_INVALID, "Invalid socket path: %s", path);
+		}
+
+		name.sun_family = AF_UNIX;
+		memcpy(name.sun_path, path, path_len);
+		size_t name_len = sizeof(name) - sizeof(name.sun_path) + path_len;
```

Comment:
I would be careful with expressions like `sizeof(x) - sizeof(x.a)` - they look weird. Even if `sun_path` is last field in struct, there is no guarantee there will be no padding for alignment purposes that will ruin this calculation. 

What is suggested in [unix(7) man](https://www.man7.org/linux/man-pages/man7/unix.7.html):

```suggestion
		size_t name_len = offsetof(struct sockaddr_un, sun_path) + path_len;
```

Reply:
I think for some reason I thought offsetof wasn't available here